pensieve/statistics.py

```diff
@@ -0,0 +1,160 @@
+from pandas import DataFrame
+import pandas
+import numpy as np
+import re
+
+from pensieve.pre_treatment import PreTreatment, RemoveNulls
+
+
+@attr.s(auto_attribs=True)
+class StatisticResult:
+    """
+    Represents the resulting data after applying a statistic transformation
+    to metric data.
+    """
+
+    metric: str
+    statistic: str
+    parameter: Decimal
+    label: str
+    ci_width: Optional[float] = 0.0
+    point: Optional[float] = 0.0
+    lower: Optional[float] = 0.0
+    upper: Optional[float] = 0.0
+
+
+@attr.s(auto_attribs=True)
+class StatisticResultCollection:
+    """
+    Represents a set of statistics result data.
+    """
+
+    data: List[StatisticResult] = []
+
+    def save_to_bigquery(self, client, destination_table, append=True):
+        """Stores the data to a BigQuery table with a defined schema."""
+
+        job_config = bigquery.LoadJobConfig()
+        job_config.schema = [
+            bigquery.SchemaField("metric", "STRING"),
+            bigquery.SchemaField("statistic", "STRING"),
+            bigquery.SchemaField("parameter", "NUMERIC"),
+            bigquery.SchemaField("label", "STRING"),
+            bigquery.SchemaField("ci_width", "FLOAT64"),
+            bigquery.SchemaField("point", "FLOAT64"),
+            bigquery.SchemaField("lower", "FLOAT64"),
+            bigquery.SchemaField("upper", "FLOAT64"),
+        ]
+
+        if append:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_APPEND
+        else:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_TRUNCATE
+
+        client.load_table_from_json(
+            self.to_dict()["data"], destination_table, job_config=job_config
+        )
+
+    def to_dict(self):
+        """Return statistic results as dict."""
+
+        return attr.asdict(self)
+
+
+@attr.s(auto_attribs=True)
+class Statistic(ABC):
+    """
+    Abstract representation of a statistic.
+
+    A statistic is a transformation that accepts a table of per-client aggregates and
+    returns a table representing a summary of the aggregates with respect to the branches
+    of the experiment.
+    """
+
+    pre_treatments: List[PreTreatment]
+
+    @classmethod
+    def name(cls):
+        """Return snake-cased name of the statistic."""
+        return re.sub(r"(?<!^)(?=[A-Z])", "_", cls.__name__).lower()
+
+    def apply(self, df: DataFrame, metric: str) -> "StatisticResultCollection":
+        """
+        Run statistic on data provided by a DataFrame and return a collection
+        of statistic results.
+        """
+
+        data = df
+        for pre_treatment in self.pre_treatments:
+            data = pre_treatment.apply(data)
+
+        col = StatisticResultCollection([])
+
+        if metric in df:
+            col.data += self.transform(df, metric).data
+
+        return col
+
+    def transform(self, df: DataFrame, metric: str) -> "StatisticResultCollection":
+        return NotImplemented
```

Comment:
Is it true that for a specific Statistic the same pre-treatments should get applied? Or can this be different and should this be configurable? If the latter is the case then `pre_treatments` should moved out of this class and moved to what is currently called `AnalysisStep` (which represents the combination of a metric and a statistic). Then having separate `apply` and `transform` methods would not be necessary here.

Reply:
> If the latter is the case then pre_treatments should moved out of this class and moved to what is currently called AnalysisStep (which represents the combination of a metric and a statistic). Then having separate apply and transform methods would not be necessary here.

Sorry for not answering this sooner. I think the pretreatments depend on the metric, and the pretreatments need to be properties of the Summary.